### PR TITLE
Ant colony optimization

### PR DESCRIPTION
**Pull request by tuongTr**
:new: :bomb: :mortar_board: :memo: :robot: :lipstick: :bug: :telescope: :art: 

### Additional Notes

This is my second attempt for a pull request. Can you check if there are any problem? Thank you!

### CODE REVIEW

**Comment by tuongTr** [comment]
good?

File: site/student-lectures/ant-colony/ant-colony-optimization.rst [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+------------
+* Nature has evolved over millions of years, fine-tuning solutions to survival challenges. These solutions, often embedded in the behaviors of organisms, have inspired the creation of algorithms that mimic natural processes.
+
+* To solve complicated problems, nature-inspired algorithms harness the power of evolutionary processes, swarm intelligence, and self-organization.
+
+    .. figure:: Safari_ants.png
+        :align: center
+    Ant behavior was the inspiration for the metaheuristic optimization technique.
+    By Mehmet Karatay - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=2179109
+
```
Comment (jameshughes89) [inline site/student-lectures/ant-colony/ant-colony-optimization.rst:14]: if this is a caption, I think you need another new line and it needs to be indented. Further, set a `:target:` to the figure to be the URL.

File: site/student-lectures/ant-colony/ant-colony-optimization.rst [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+* **Pheromone Communication**: As ants discover food, they return to the nest while depositing a chemical substance called pheromone on the ground. This forms a trail that other ants can follow.
+
+* **Positive Feedback**: Shorter paths accumulate pheromones more quickly than longer paths due to the faster travel time. This creates a positive feedback loop where shorter paths attract more ants.
+
+* **Exploitation**: Over time, the pheromone concentration on shorter paths increases, making them more attractive to other ants. This results in a collective focus on the most efficient paths.
+
+    .. figure:: Artificial_ants.png
+        :align: center
+    By Jean-Baptiste Waldner - Source : &quot;Nanocomputers and Swarm Intelligence&quot;, Jean-Baptiste Waldner, John Wiley &amp; Sons, 2008., CC BY 3.0, https://commons.wikimedia.org/w/index.php?curid=3435721
```
Comment (jameshughes89) [inline site/student-lectures/ant-colony/ant-colony-optimization.rst:32]: same as above.

File: site/student-lectures/ant-colony/ant-colony-optimization.rst [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+5. **Ant Movement:**
+   - Ants construct solutions by iteratively selecting solution components based on a probabilistic rule that considers both pheromone levels and a heuristic measure.
+
+6. **Solution Evaluation:**
+   - Evaluate the quality of the solutions constructed by ants.
+
+7. **Pheromone Update:**
+   - Update pheromone levels based on the quality of the solutions. Good solutions receive higher pheromone deposits.
+
+8. **Iteration:**
```
Comment (jameshughes89) [inline site/student-lectures/ant-colony/ant-colony-optimization.rst:96]: All numbers can be replaced with `#.`s. Sphinx will then automatically number everything.

File: site/student-lectures/ant-colony/ant-colony-optimization.rst [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+   - Repeat the ant movement, solution evaluation, and pheromone update steps for a specified number of iterations or until a termination criterion is met.
+
+    .. figure:: Aco_shortpath.svs.png
+        :align: center
+    By Nojhan - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=821076
```
Comment (jameshughes89) [inline site/student-lectures/ant-colony/ant-colony-optimization.rst:101]: Same as above.

File: site/student-lectures/ant-colony/ant-colony-optimization.rst [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+
+Applications of ACO
+-------------------
+
+* Ant Colony Optimization (ACO) algorithms have found widespread application in solving diverse combinatorial optimization problems. From quadratic assignment and protein folding to vehicle routing, these algorithms and their derivatives have been adapted for dynamic problems, real variables, stochastic scenarios, multi-targets, and parallel implementations.
+
+\
```
Comment (jameshughes89) [inline site/student-lectures/ant-colony/ant-colony-optimization.rst:109]: ?

File: site/student-lectures/ant-colony/ant-colony-optimization.rst [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+* Traveling Salesman Problem (TSP):
+
+    .. figure:: 1920px-Aco_TSP.svg.png
+        :align: center
+
+    1) An ant choose a path among other, and lay a pheromonal trail on it.
+    2) All the ants are travelling some paths, laying a trail proportionnal to the quality of the solution.
+    3) Each edge of the best path is more reinforced than others.
+    4) Evaporation ensures that the bad solutions disappear.
```
Comment (jameshughes89) [inline site/student-lectures/ant-colony/ant-colony-optimization.rst:126]: Same as above

File: site/student-lectures/ant-colony/ant-colony-optimization.rst [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+4. **Solution Construction Heuristics:**
+   - ACO provides a solution construction mechanism that incrementally builds solutions, leveraging both pheromone information and heuristic knowledge.
+
+5. **Scalability:**
+   - ACO is scalable and can handle large problem instances by distributing the exploration across multiple agents.
+
+6. **Applicability to Various Domains:**
+   - ACO has been successfully applied to a wide range of problems, including the Traveling Salesman Problem (TSP), Job Scheduling, Network Routing, and more.
+
+7. **Natural Robustness:**
```
Comment (jameshughes89) [inline site/student-lectures/ant-colony/ant-colony-optimization.rst:155]: Same as above

File: site/student-lectures/ant-colony/ant-colony-optimization.rst [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+   - ACO may require significant memory and storage resources, particularly when dealing with large problem instances or a large number of iterations.
+
+4. **Dependency on Heuristics:**
+   - The success of ACO often relies on the availability of effective heuristic information, and the choice of heuristics can impact the algorithm's performance.
+
+5. **Limited Handling of Continuous Spaces:**
+   - ACO is inherently designed for discrete problems, and adapting it to continuous optimization spaces can be non-trivial.
+
+6. **Local Optima:**
+   - ACO may struggle in scenarios with deceptive landscapes, where local optima mislead the algorithm away from the global optimum.
```
Comment (jameshughes89) [inline site/student-lectures/ant-colony/ant-colony-optimization.rst:176]: Same as above

**Review by jameshughes89 — CHANGES_REQUESTED** [review]
I made some small comments. 

1. Perhaps provide a more concrete example?
2. Did you build this locally to check your code? Judging from the code, it seems that maybe you have not.

File: site/student-lectures/ant-colony/ant-colony-optimization.rst (hunk 1 above)
Comment (tuongTr) [reply site/student-lectures/ant-colony/ant-colony-optimization.rst:14]: Thanks, I fixed it.

File: site/student-lectures/ant-colony/ant-colony-optimization.rst (hunk 2 above)
Comment (tuongTr) [reply site/student-lectures/ant-colony/ant-colony-optimization.rst:32]: Thanks, I fixed it.

File: site/student-lectures/ant-colony/ant-colony-optimization.rst (hunk 4 above)
Comment (tuongTr) [reply site/student-lectures/ant-colony/ant-colony-optimization.rst:101]: Thanks, I fixed it.

File: site/student-lectures/ant-colony/ant-colony-optimization.rst (hunk 5 above)
Comment (tuongTr) [reply site/student-lectures/ant-colony/ant-colony-optimization.rst:109]: Thanks, I fixed it.

File: site/student-lectures/ant-colony/ant-colony-optimization.rst (hunk 6 above)
Comment (tuongTr) [reply site/student-lectures/ant-colony/ant-colony-optimization.rst:126]: Thanks, I fixed it.

File: site/student-lectures/ant-colony/ant-colony-optimization.rst (hunk 7 above)
Comment (tuongTr) [reply site/student-lectures/ant-colony/ant-colony-optimization.rst:155]: Thanks, I fixed it.

File: site/student-lectures/ant-colony/ant-colony-optimization.rst (hunk 8 above)
Comment (tuongTr) [reply site/student-lectures/ant-colony/ant-colony-optimization.rst:176]: Thanks, I fixed it.

**Comment by tuongTr** [comment]
I add two real life example, can you check if there are anything else to add? Thank you.

I try to run the code locally, but mine keep crashing, it said permission denied.

**Review by jameshughes89 — CHANGES_REQUESTED** [review]
You need to confirm that it renders correctly before I will accept the PR. You can check the action artifact [here](https://github.com/jameshughes89/cs4XX-EvolutionaryComputation/actions/runs/6895639512)

File: site/student-lectures/ant-colony/ant-colony-optimization.rst [hunk 9]
```diff
@@ -9,8 +9,10 @@ Introduction
 
     .. figure:: Safari_ants.png
         :align: center
-    Ant behavior was the inspiration for the metaheuristic optimization technique.
-    By Mehmet Karatay - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=2179109
+
+        Ant behavior was the inspiration for the metaheuristic optimization technique. By Mehmet Karatay - Own work, CC BY-SA 3.0
+        :target: https://commons.wikimedia.org/w/index.php?curid=2179109
```
[the remark this answers is not in the document]
Comment (tuongTr) [reply site/student-lectures/ant-colony/ant-colony-optimization.rst]: thanks, I fixed it, everything seems good to me

**Comment by tuongTr** [comment]
It renders good

**Review by jameshughes89 — CHANGES_REQUESTED** [review]
All that's left to do is add your page to the index.

**Comment by tuongTr** [comment]
so how do I do it?

**Comment by jameshughes89** [comment]
> so how do I do it?

Just look at the current index and see what everyone else did. Basically copy that.

**Comment by tuongTr** [comment]
I updated my index.rst

**Review by jameshughes89 — APPROVED** [review]

**Pull request merged**